src/unix/signal.c

```diff
@@ -228,16 +226,35 @@ static void deliver_signal(sigstate ss, struct siginfo *info)
     heap h = heap_general(get_kernel_heaps());
     int sig = info->si_signo;
 
-    /* check if we can post */
-    if (sig < RT_SIG_START && sigstate_is_pending(ss, sig)) {
-        /* Standard signal already posted. Unless a particular signal
-           would allow the updating of posted siginfo, just return here. */
-        sig_debug("already posted; ignore\n");
-        return;
+    /* Special handling for pending signals */
```

Comment:
Right, perhaps the signal delivery can return the queued_signal as a handle and a call can be added to update it. I'll take a crack at that before merging.

Reply:
On second thought, this would then require that a signal dequeue call back into the timer code to nullify the reference. If we were going to do that, we may as well never push overrun updates and instead add special handling for SI_TIMER on dequeue to call into the timer code for the latest value of overrun. I believe this is what Linux does. That's kind of kludgey, too, but it would be less overhead per timer expiry.

I guess I'm inclined to punt for now if there isn't a correctness issue. I also kind of wonder how likely it is that this search would be for N > 1. If a program is combining SI_TIMER and other queueable events on the same rt signal (signo >= 32), then yes it does mean unnecessary searching.

I guess if we actually see very high rate posix timers in real-world use, we can revisit this. Maybe a little mechanism to install a general closure callback to update a siginfo before finally dispatching to a thread would be a nice thing to have?

Reply:
i wouldn't block this if it's functional - can always open an issue and track future work desired